caproto/ioc_examples/lakeshore.py

```diff
@@ -0,0 +1,390 @@
+        self._run = True
+        threading.Thread(target=self._simulate).start()
+
+    @property
+    def temperature(self):
+        return self.energy / self.thermal_mass
+
+    @property
+    def heater_power(self):
+        return self._heater_power
+
+    @heater_power.setter
+    def heater_power(self, value):
+        self._heater_power = value
+
+    def set_heater_power(self, value):
+        self._heater_power = value
+
+    def stop(self):
+        self._run = False
+
+    def _cooling(self):
+        return -1 * self.cooling_constant * (self.temperature - self.ambient_temp)
+
+    def _heating(self):
+        return self.heater_power
+
+    def _simulate(self):
+        while self._run:
+            now = time.time()
+            time_delta = now - self.time
+            self.time = now
+            self.energy += self._cooling() * time_delta
+            self.energy += self._heating() * time_delta
+            time.sleep(0.1)
+
+
+class PIDController(PID):
+    """
+    General purpose PID controller that supports ramping.
+
+    Parameters
+    ----------
+    get_feedback: callable
+        A function that returns the feedback value of the system.
+    set_output: callable
+        A function that sets the output value of the system.
+    ramp_rate: int, float
+        The rate that the setpoint should ramp at.
+    setpoint: int, float
+
+    Example
+    -------
+    import matplotlib.pyplot as plt
+    from matplotlib.animation import FuncAnimation
+    from caproto.ioc_examples.lakeshore import (PIDController,
+    ThermalMaterial)
+
+    sample = ThermalMaterial()
+    setpoint = 150
+
+    temperature_controller = PIDController(
+        lambda: sample.temperature,
+        sample.set_heater_power,
+        Kp=1,
+        Ki=0.1,
+        Kd=0.05,
+        ramp_rate=1,
+        setpoint=setpoint,
+    )
+
+    def update(i):
+        x_values = temperature_controller.history['timestamp']
+        plt.cla()
+        plt.plot(x_values, temperature_controller.history['feedback'])
+        plt.plot(x_values, temperature_controller.history['output'])
+        plt.plot(x_values, temperature_controller.history['setpoint'])
+        plt.xlabel('time')
+        plt.ylabel('temperature')
+        plt.title('Temperature Controller')
+        plt.gcf().autofmt_xdate()
+        plt.tight_layout()
+
+    ani = FuncAnimation(plt.gcf(), update, 1000)
+    plt.tight_layout()
+    plt.show(block=False)
+    """
+
+    def __init__(
+        self, get_feedback, set_output, ramp_rate=1, setpoint=150, *args, **kwargs
+    ):
+        self._setpoint = setpoint
+        self.ramp_rate = ramp_rate
+        super().__init__(setpoint=setpoint, *args, **kwargs)
+        self._get_feedback = get_feedback
+        self._set_output = set_output
+        self._feedback = None
+        self._output = None
+        self._run = True
+        self._ramping = False
+        self.history = {"output": [], "feedback": [], "setpoint": [], "timestamp": []}
```

Comment:
As a general rule, servers should not hold data structures that grow without bound. Would it work to make these lists `collections.deque` objects with a reasonable `maxlen`?